File: Coding-Interviews/20-表示数值的字符串.cpp
#include <string>
using namespace std;
class Solution
{
public:
    bool isNumber(string s)
    {
        int cur = 0;
        while (s[cur] == ' ')
            cur++;
        if (s[cur] == '+' || s[cur] == '-')
            cur++;
        if ('0' <= s[cur] && s[cur] <= '9') {
            cur++;
            while ('0' <= s[cur] && s[cur] <= '9')
                cur++;
            if (s[cur] == '.')
                cur++;
            while ('0' <= s[cur] && s[cur] <= '9')
                cur++;
        }
        else if (s[cur] == '.') {
            cur++;
            if (s[cur] < '0' || s[cur] > '9')
                return false;
            while ('0' <= s[cur] && s[cur] <= '9')
                cur++;
        }
        else
            return false;
        if (s[cur] == 'e' || s[cur] == 'E') {
            cur++;
            if (s[cur] == '+' || s[cur] == '-')
                cur++;
            if (s[cur] < '0' || s[cur] > '9')
                return false;
            while ('0' <= s[cur] && s[cur] <= '9')
                cur++;
        }
        while (s[cur] == ' ')
            cur++;
        return cur == s.size() ? true : false;
    }
};

File: Coding-Interviews/20-表示数值的字符串.py
class Solution:
    def isNumber(self, s: str) -> bool:
        idx = 0
        while idx < len(s) and s[idx] == ' ':
            idx += 1
        if idx < len(s) and (s[idx] == '+' or s[idx] == '-'):
            idx += 1
        if idx < len(s) and '0' <= s[idx] <= '9':
            idx += 1
            while idx < len(s) and '0' <= s[idx] <= '9':
                idx += 1
            if idx < len(s) and s[idx] == '.':
                idx += 1
                while idx < len(s) and '0' <= s[idx] <= '9':
                    idx += 1
        elif idx < len(s) and s[idx] == '.':
            idx += 1
            if idx < len(s) and '0' <= s[idx] <= '9':
                idx += 1
            else:
                return False
            while idx < len(s) and '0' <= s[idx] <= '9':
                idx += 1
        else:
            return False
        if idx < len(s) and (s[idx] == 'e' or s[idx] == 'E'):
            idx += 1
            if idx < len(s) and (s[idx] == '-' or s[idx] == '+'):
                idx += 1
            if idx < len(s) and '0' <= s[idx] <= '9':
                idx += 1
                while idx < len(s) and '0' <= s[idx] <= '9':
                    idx += 1
            else:
                return False
        while idx < len(s) and s[idx] == ' ':
            idx += 1
        return len(s) == idx
